### DISCUSSION TITLE
Ethernet mod possible?

### DISCUSSION THREAD

**Discussion by spali** [emporia-vue-local/esphome#67] [Q&A] [answered]
I'm interested in buying one of these especially because of this project.
But because I have bad WiFi in the basement, I wanted to ask if anyone tried a Ethernet shield mod?
Or at least, how hard would it be to get the right pins soldered?
For Ethernet, I assume I need either I2C or SPI access.
Which one would be easier in this particular case for the vue?
I'm more concerned about the Hardware part as I'm bad at soldering everything that's smaller than standard pin headers. So hope it could be done without solder directly to the esp pins.

**Comment by flaviut** [comment, marked as answer]
First off, I want to note that having low-voltage wiring like ethernet in a high-voltage box like your electrical panel is forbidden by the electrical code, and also that a physical modification like this would remove the UL listing on the Vue.

An alternative solution might be to bring ethernet to your basement, and attach a WiFi repeater to it.

Anyway, that said, you would definitely end up having to solder to the ESP32 directly. The necessary pins are not broken out. Looks like these shields typically use SPI, so 6 pins to solder probably?

**Reply by spali** [reply]
> First off, I want to note that having low-voltage wiring like ethernet in a high-voltage box like your electrical panel is forbidden by the electrical code, and also that a physical modification like this would remove the UL listing on the Vue.

That's right, but I don't care 😄 

> An alternative solution might be to bring ethernet to your basement, and attach a WiFi repeater to it.

I have ethernet and even wifi in the basement, but I prefer the reliability of a cable.
 
> Anyway, that said, you would definitely end up having to solder to the ESP32 directly. The necessary pins are not broken out.

Damn... but maybe I still give it a try. 

> Looks like these shields typically use SPI, so 6 pins to solder probably?

5 (4 for SPI including CS and 1 for interrupt) if you skip the reset pin which is not required.



But thanks for the feedback, I think I order one and see how it goes.

**Reply by spali** [reply]
Actually you are correct, missed the power pins.. so 6 without interrupt.

**Reply by oliv3r** [reply]
I was thinking the same; not thinking of code of course. What you'd ideally want, is to have is a ESP32-WROVER module (from pinout pov) that instead of wifi, just uses ethernet (the ESP32 has native ethernet).

I don't know if such a module exists, creating one should be fairly trivial (the hard part is the wifi radio design). The PCB would have to re-route some of those IO's potentially to a different ESP gpio though; if the EMAC pins are used for something else. The upside is that the software would be almost the same.

Btw, as to 'code' If the box is mounted outside of the panel, then technically the vue is not inside the electrical box anymore, just the CT wires coming out ... some food for thought ...

**Reply by spali** [reply]
I thought about a W5500, which just needs SPI and an interrupt pin at least. Maybe less inversive and less potential to mess something else up. Especially as I read that vue2 already heavily uses the i2c bus.

Had some luck in the last days to get such a module working in esphome with an ESP32S3 for a usb host to ethernet bridge.

Regarding the "compliance". I'm not an expert, but here in Switzerland, you can put everything in the same distribution box, as long as the isolation between low and height voltage meets the specs. 
I mean, I don't care the official specs, but I don't want to fry my ethernet network from height voltage. Assuming the vue2 has already a good enough isolation between height voltage and low voltage part, the only I can do to improve this is to take an isolated network board. I think I have seen some of them in the wild. Worst case using an optocoupler for the network cable will probably do it too.

I ordered a vue2 anyway now, so I will see if it's worth the effort to get it on cable.

**Reply by oliv3r** [reply]
I think the rules are the same here in NL ;) anyway, the VUE has good isolation, so there shouldn't be much of a concern. And we do have ethernet enabled devices that sit on DIN rails ... and ethernet by design has isolation transformers :)

**Reply by crlogic** [reply]
> I think the rules are the same here in NL ;) anyway, the VUE has good isolation, so there shouldn't be much of a concern. And we do have ethernet enabled devices that sit on DIN rails ... and ethernet by design has isolation transformers :)

Physically disconnect i2c pins from the Vue2 and wire (+ ground?) to an olimex-esp32-poe?  Maybe even re-use the external antenna connector for the i2c so you can have the Olimex outside the panel?

```
i2c:
  sda: 21
  scl: 22
```

**Reply by deggle** [reply]
Did anyone get anywhere with this?

I've gone with the Vue product as it's a great price for a 3ph monitor with 16 additional clamps, and I'll need 4-5 of them (yes, it's quite a big property, with multiple distribution panels). I'm happy with Wi-Fi but would be very interested in adding ethernet. Also not too concerned about the isolation/regs/code (in that, I understand the reasoning behind the code/regulations, but in this case I feel the risks are low and are managed).

The soldering doesn't phase me, so happy to connect directly to the ESP pins if needs be. However, I also liked the suggestion of using an external POE ESP32 and just passing the I2C to the Vue - if indeed that's all it needs - can someone confirm that?

Has anyone reverse engineered the Vue board, or at least traced the ESP32 pinout to see which pins/IO are used for what?

**Reply by crlogic** [reply]
Here you go.. ;)

https://github.com/emporia-vue-local/esphome

**Reply by deggle** [reply]
Sorry, yes I'm aware (this is the discussion pages for that project) - I'm specifically interested to see if anyone has managed to implement ethernet connectivity. It's easy to do the configuration via ESPHome so that's not the issue, the question is more around the hardware - exactly as per the thread above (from a year ago, so not unreasonably I'm asking if anyone has achieved it).

**Reply by clowrey** [reply]
If anyone still wants an Ethernet version I would recommend this board (haven't tried yet).. but it's meant for ESPhome too. I have other Kincony boards and they are good.
KC868-M16v2
https://s.click.aliexpress.com/e/_mL3uVJ2

**Reply by deggle** [reply]
Ooh, that does look interesting, thank you for that!
Shame it has less channels as the Vue has 16 + 3 incoming clamps (for 3ph) whereas this board only has a total of 16 - is that right?
However, it ticks a lot of boxes with LAN, plus POE and other nice to haves like temp etc.
Will grab one I think.

**Reply by clowrey** [reply]
Be aware that the current clamp plugs are the larger 3.5mm so won't fit the small Vue clamps.. The Vue is still vastly lower cost if you figure in the clamp cost so im still using all Vue here.

**Reply by deggle** [reply]
Yes, that's a really good point. Also the Vue is well enclosed/self contained. I'm so on the fence, but perhaps initially will stick with Vue. I'm very much _wired first_ but in reality wireless connections tend to be plenty stable enough for this type of thing, and it's not the type of data where a small gap is critical anyway. Plus it's probably quite possible to add the ethernet circuit into the Vue if I really tried. Hmm.

**Reply by catatung** [reply]
I am the same way, I always 100% prefer wired connections.  I have plans to get an RP-SMA to RP-SMA RF cable with some SMA attenuators and create a link between the antenna port of the VUE2 and a small router.  Google for "GL.inet" - Configuring this correctly will allow the LAN port to plug into your LAN and any wireless devices connecting will be on the same subnet and thus be available to your entire network, while drastically (or even ideally eliminating) wifi reception for any other devices since an antenna is not being used.  The only concept of antenna here would be the leakage within the SMA cable, as far as I know.

I have not done this yet, so I do not have details yet on how well it works, but I will report back once I have time to do this.  I am estimating to start around 35db attenuation.  I intend connection to be like this:

EmporiaVue2 RP-SMA port --> 35db attenuator -> gl.inet router antenna RP-SMA port

If someone else here likes the idea and has the time now, you might beat me to it and it would be nice to hear your findings in a reply post.

**Reply by deggle** [reply]
That's an interesting solution - it still _feels_ like wireless but you've closed the loop so I guess in theory it should be much more stable.

May also be worth looking at MicroTik for the mini router. They do some _very_ good value bits, though unfortunately the ones that would be ideal (mAP & hAP) don't appear to have SMA connections. The mAP lite is tiny, PoE and runs full RouterOS but it's got an internal antenna (though perhaps it has an internal U.Fl connector - I have one, I will check).

**Reply by deggle** [reply]
I opened up my mAP lite, and nope, the antennas are on-PCB tracks with no connections (could possibly do a trace cut and solder job on them). However, while not quite as fool proof as having the direct SMA cable linking, I do wonder if a mAP besite the Vue2 with a hidden SSID network would give almost the reliability desired. Sure, it won't be immune from external noise (on the same channel) but it would be independent from the wider wireless network. Something to think about.

**Reply by catatung** [reply]
Yes lots of devices like that internal design of PCB copper tracks for an antenna.  I've already tested my idea (from a network configuration perspective) with a GL.Inet mini router, and it works well.  I actually have it set up now with a regular wifi and regular antennas.  I will eventually have time to test my idea with an SMA cable and attenuator.

The model I have used is: https://www.ebay.com/itm/314867315252

Short of that I had another idea of a small faraday cage and putting both inside of it, but this seems much cheaper and simpler and so I will be eventually looking into this method.

My setup also has the Emporia VUE2 outside of the electrical panel -- I've got just the CT clamps inside the panel and all exiting out of a knock out at the bottom of the panel (with a grommet of course).  Then I mounted the VUE2 below the panel and ran the wires to it.

**Comment by oliv3r** [comment]
Btw, @crlogic i just found these (feels a bit like blast from the past) ;)

http://www.hi-flying.com/index.php?route=product/search&search=eport the linux module looks cool; but even the m3 could be super useful :) wonder what they cost though; trying to hunt down ebay/ali express for a couple of these now :)

**Comment by oliv3r** [comment]
https://github.com/emporia-vue-local/esphome/discussions/264

looks like gen3 comes with ethernet out of the box.
